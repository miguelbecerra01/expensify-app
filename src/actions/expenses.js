//https://www.npmjs.com/package/uuid
import uuid from 'uuid';
import database from '../firebase/firebase';

//Actions 

//steps-> Syncronous calls
//component calls action generator
//action generator returns object
//component dispatches object
//redux store changes

//steps -> Asyncronous calls (save to firebase)
//component calls action generator
//action generator returns function
//component dispatches function (?) using a redux-module redux-thunk
//(a thunk is a subroutine used to inject an additional calculation into another subroutine)
//function runs (has the ability to dispatch other actions and do whatever it wants)


// ADD_EXPENSE
// export const addExpense = (
//     { description = '',
//         note = '',
//         amount = 0,
//         createdAt = 0
//     } = {}
// ) => ({
//     type: 'ADD_EXPENSE',
//     expense: {
//         id: uuid(),
//         description,
//         note,
//         amount,
//         createdAt
//     }
// });

export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});


export const startAddExpense = (expenseData = {}) => {
    //will work only because we added the redux middleware using redux-thunk
    return (dispatch) => {
        //extract and create the parameters from the object expenseData -> destructuring
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;

        const expense = { description, note, amount, createdAt };

        //with return we add the hability to make promises chaining
        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
};


// REMOVE_EXPENSE 
//with parentesis after the arrow function means that returns the object
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

