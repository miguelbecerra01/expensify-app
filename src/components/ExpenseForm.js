import React from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';

//its a class because we need to use component state
export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: false,
            submitType: props.submitType === 'Add' ? 'Add' : 'Edit'
        };
    }
    // state = {
    //     description: '',
    //     note: '',
    //     amount: '',
    //     createdAt: moment(),
    //     calendarFocused: false,
    //     error: false
    // };
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => {
            this.state.error = !description;
            return {
                description
            }
        });
    };
    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => {
                this.state.error = !amount;
                return {
                    amount
                };
            });
        }
    };
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };
    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };
    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: 'Please provide description and amount.' }));
        } else {
            this.setState(() => ({ error: '' }));
            //call to the parent method in AddExpensePage, returning the value
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                //valueOf is from momentjs, because createdAt is a moment object
                //with valueOf returns in miliseconds 
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    };
    render() {
        return (
            <div>

                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        name="description"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange} />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange} />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        firstDayOfWeek={1}
                        small={true}
                        isOutsideRange={() => false} />
                    <textarea
                        placeholder="Add a note for your expense (optional)"
                        onChange={this.onNoteChange}
                        value={this.state.note}
                    >

                    </textarea>
                    <button >{this.state.submitType} Expense</button>
                </form>
            </div>
        );
    }
}