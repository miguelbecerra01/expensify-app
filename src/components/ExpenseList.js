import React from 'react';
//connect -> connect the component with the redux store
import { connect } from 'react-redux';


const ExpenseList = (props) => (
    <div>
        <h1>Expense List</h1>
        {props.expenses.length}
        <ul>
            <p>Text filter: {props.filters.text}</p> 
            {props.expenses.map((expense, index) => (
                <li key={index}>{expense.description}</li>
            ))}
        </ul>
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: state.expenses,
        filters: state.filters
    };
};

//Higher order component -> enhance the ExpenseList component adding the store of redux
//connect returns a function  and we need to set the component to propagate the store values in the props of the component
//the api needs to send the components in the second parentesis connect(function)(componentToPropagateStore)
//const ConnectedExpenseList = connect((state) => {
export default connect(mapStateToProps)(ExpenseList);

//export default ConnectedExpenseList;