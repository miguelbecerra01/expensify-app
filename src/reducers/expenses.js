//Expenses Reducer
const expensesReducerDefaultState = [];
const expensesReducer = (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            //return state.concat(action.expense);
            return [...state, action.expense];
        case 'REMOVE_EXPENSE':
            //we need to destructure the array of objects in order to filter by id, otherwise it wont work    
            return state.filter(({ id }) => id !== action.id);
        case 'EDIT_EXPENSE':
            //loop througth every item on the array
            return state.map((expense) => {
                if (expense.id === action.id) {
                    //spread operator for override the object with new values
                    //const expenseEdited = {...expense,...action.updates};
                    //return expenseEdited;
                    return {
                        ...expense,
                        ...action.updates
                    }
                } else {
                    return expense;
                }
            });
        case 'SET_EXPENSES':
            return action.expenses;
        default:
            return state;
    }
};


export default expensesReducer;