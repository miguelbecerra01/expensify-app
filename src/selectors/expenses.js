//Get visible expenses

const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {

    return expenses.filter((expense) => {
        const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
        const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
        const textMatch = (expense.description.toLowerCase()).includes(text.toLowerCase());

        //if is a match
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy === 'date') {
            //sort descending
            return a.createdAt < b.createdAt ? 1 : -1;
        }
            //sort descending
        if (sortBy === 'amount') {
            //if is true b comes first if is false a comes first
            return a.amount < b.amount ? 1 : -1;
        }
    });
};

export default getVisibleExpenses;