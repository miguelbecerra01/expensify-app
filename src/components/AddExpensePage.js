import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            isLoading: false
        }
    };
    onSubmit = (expense) => {
        this.setState(() => ({ isLoading: true }));

        new Promise((resolve, reject) => {
            resolve(this.props.startAddExpense(expense));
        }).then(() => {
            this.props.history.push('/');
        });
    };
    render() {
        return (
            <div>
                <h1>Add Expense Pages</h1>
                {this.state.isLoading ?
                    <div><img src="/images/loading.gif" width="100px" height="100px" /></div> :
                    <ExpenseForm submitType={'Add'} isLoading={this.isLoading} onSubmit={this.onSubmit} />
                }

            </div>
        );
    }
}


//https://react-redux.js.org/using-react-redux/connect-mapdispatch#connect-dispatching-actions-with-mapdispatchtoprops
//Its a way to return your dispatch functions allowing you to abstract them 
//away of the component itseft
const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
