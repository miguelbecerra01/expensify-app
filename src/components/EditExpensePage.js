import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            isLoading: false
        }
    };
    onSubmit = (expense) => {
        this.setState(() => ({ isLoading: true }));

        new Promise((resolve, reject) => {
            resolve(this.props.startEditExpense(this.props.expense.id, expense));
        }).then(() => {
            this.props.history.push('/');
        });
    };
    onRemove = () => {
        this.setState(() => ({ isLoading: true }));
        new Promise((resolve, reject) => {
            resolve(this.props.startRemoveExpense({ id: this.props.expense.id }));
        }).then(() => {
            this.props.history.push('/');
        });
    };
    render() {
        return (
            <div>
                <h1>Edit Expense</h1>
                {this.state.isLoading ?
                    <div><img src="/images/loading.gif" width="100px" height="100px" /></div> :
                    <div>
                        <ExpenseForm
                            submitType={'Edit'}
                            expense={this.props.expense}
                            onSubmit={this.onSubmit} />
                        <button onClick={this.onRemove}>Remove</button>
                    </div>
                }
            </div>
        );
    }
}

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => (expense.id === props.match.params.id))
});

const mapDispatchToProps = (dispatch) => ({
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);