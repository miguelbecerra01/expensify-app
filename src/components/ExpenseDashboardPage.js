import React from 'react';

import * as moment from 'moment/moment';
moment.locale('es');

import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseSummary from './ExpenseSummary';

export class ExpenseDashboardPage extends React.Component {

    render() {
        return (
            <div>
                <hr></hr>
                <ExpenseSummary />
                <ExpenseListFilters />
                <ExpenseList />
            </div>
        );
    }
}


export default ExpenseDashboardPage;