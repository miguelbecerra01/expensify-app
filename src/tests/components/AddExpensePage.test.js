import React from 'react';
import { shallow } from 'enzyme';
import { AddExpensePage } from '../../components/AddExpensePage';
import expenses from '../fixtures/expenses';

//https://jestjs.io/docs/en/api#beforeeachfn-timeout


//Runs a function before each of the tests in this file runs. 
let startAddExpense, historySpy, wrapper;

beforeEach(() => {
    startAddExpense = jest.fn(); //tells that is a jest function to mock
    historySpy = { push: jest.fn() };
    wrapper = shallow(<AddExpensePage startAddExpense={startAddExpense} history={historySpy} />);
});

test('should render AddExpensePage correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);

    expect(startAddExpense).toHaveBeenLastCalledWith(expenses[1]);
    //expect(historySpy.push).toHaveBeenLastCalledWith('/');
});