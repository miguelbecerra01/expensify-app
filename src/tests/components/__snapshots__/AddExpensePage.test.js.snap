// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddExpensePage correctly 1`] = `
<div>
  <h1>
    Add Expense Pages
  </h1>
  <ExpenseForm
    onSubmit={[Function]}
    submitType="Add"
  />
</div>
`;
